packages/abi-typegen/src/abi/abi.ts

```diff
@@ -0,0 +1,133 @@
+import {
+  DTS_TEMPLATE_ENCODER,
+  DTS_TEMPLATE_DECODER,
+  DTS_TEMPLATE_STRUCT,
+  DTS_TEMPLATE_ENUM,
+  DTS_TEMPLATE,
+} from '../templates/dts';
+import { FACTORY_TEMPLATE } from '../templates/factory';
+import { normalizeName } from '../utils/normalize';
+
+import { parseFunctions } from './functions';
+import type { IAbi } from './interfaces/IAbi';
+import type { IFunction } from './interfaces/IFunction';
+import type { IType } from './interfaces/IType';
+import { parseTypes } from './types';
+import type { EnumType } from './types/EnumType';
+import type { StructType } from './types/StructType';
+
+export class Abi {
+  public name: string;
+  public filepath: string;
+  public outputDir: string;
+
+  public dtsFilepath: string;
+  public factoryFilepath: string;
+
+  public usesEnum: boolean;
+
+  public rawContents: IAbi;
+  public types: IType[] = [];
+  public functions: IFunction[] = [];
+
+  constructor(params: { filepath: string; outputDir: string; rawContents: IAbi }) {
+    const { filepath, outputDir, rawContents } = params;
+
+    const abiNameRegex = /([^/]+)-abi\.json$/m;
+    const abiName = filepath.match(abiNameRegex);
+
+    if (!abiName || abiName.length === 0) {
+      throw new Error(`Could not parse name from abi file: ${filepath}`);
+    }
+
+    const name = normalizeName(abiName[1]);
+
+    this.dtsFilepath = `${outputDir}/${name}Abi.d.ts`;
+    this.factoryFilepath = `${outputDir}/factories/${name}__factory.d.ts`;
+
+    this.name = name;
+    this.filepath = filepath;
+    this.rawContents = rawContents;
+    this.outputDir = outputDir;
+
+    const { types, functions } = this.parse();
+
+    this.types = types;
+    this.functions = functions;
+    this.usesEnum = !!types.find((t) => t.name === 'enum');
+  }
+
+  parse() {
+    const { types: rawAbiTypes, functions: abiFunctions } = this.rawContents;
+
+    const types = parseTypes({ rawAbiTypes });
+    const functions = parseFunctions({ abiFunctions, types });
+
+    return {
+      types,
+      functions,
+    };
+  }
+
+  getDtsDeclaration() {
+    const { types, functions } = this;
+
+    /*
+      First we format all attributes
+    */
+    const fnsTypedefs = this.functions.map((f) => f.getDeclaration());
+
+    const fnsFragments = this.functions.map((f) => `${f.attributes.name}: FunctionFragment;`);
+
+    const encoders = functions.map((f) =>
+      DTS_TEMPLATE_ENCODER.replace('{NAME}', f.attributes.name).replace(
```

Comment:
I liked the idea of the DTS_TEMPLATE_ENCODER maybe, we should also create a wrapper function `templateEncoder(name: string, input: Input): string` this function could be inside the templates.

Reply:
Are you talking about these wrappers?
 - `encodeFunctionData`
 - `decodeFunctionData`

https://github.com/FuelLabs/fuels-ts/blob/e6e6890e9f11dc9323917101aa4e7f367ab0bb5d/packages/abi-typegen/src/templates/dts.ts#L37-L39

Reply:
Like this;
```ts
export const DTS_TEMPLATE_DECODER = `decodeFunctionData(functionFragment: '{NAME}', data: BytesLike): DecodedValue`;

export function templateDecoder(name: string, inputs: IAbiFunctionIO[]) {
    return DTS_TEMPLATE_ENCODER.replace('{NAME}', f.attributes.name).replace('{INPUT}', inputs)
```